numpy/core/src/multiarray/dtypemeta.c

```diff
@@ -398,9 +401,8 @@ static PyArray_DTypeMeta *
 string_unicode_common_dtype(PyArray_DTypeMeta *cls, PyArray_DTypeMeta *other)
 {
     assert(cls->type_num < NPY_NTYPES);
-    if (!other->legacy || other->type_num > cls->type_num ||
-        other->type_num == NPY_OBJECT) {
-        /* Let the more generic (larger type number) DType handle this */
+    if (!other->legacy || (!PyTypeNum_ISNUMBER(other->type_num) &&
+            (cls->type_num == NPY_STRING && other->type_num == NPY_UNICODE))) {
```

Comment:
The `(...)` here is unneeded and a bit confusing.

Reply:
OK, I managed to horribly break this the `&&` should be an `||`, and surprisingly none of our tests noticed, maybe because it is an asymmetry (object knows the correct result is object, just string gets it wrong).

Reply:
This needs a simple fix? Because it has been backported I need to fix it before 1.20 is released.

Reply:
Yeah, I just got the logic wrong, so it is a simple fix.  I will make a PR today.  The annoying thing is really that this oversight was able to split past the tests...